### ISSUE TITLE
Misleading error info where it should require explicit type annotations in nightly

### ISSUE THREAD

**Issue by TsuITOAR** [rust-lang/rust#118085]
### Code

```Rust
use std::io::Read;

struct Wrapper<In> {
    i: In,
}

impl<In: Read> Read for Wrapper<In> {
    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
        self.i.read(buf)
    }
}

impl<In> Read for &Wrapper<In>
where
    for<'a> &'a In: Read,
{
    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
        (&self.i).read(buf)
    }
}

trait FromReadable<S: Read>: Default {
    fn get(s: S) -> Self {
        Self::default()
    }
}

#[derive(Default)]
struct Got;

impl<S: Read> FromReadable<S> for Got {}
fn from_readable<S: Read, G: FromReadable<S>>(g: G) {}

impl<In> Wrapper<In>
where
    for<'a> &'a In: Read,
{
    fn foo(&self) {
        let got = Got::get(self);
        //from_readable::<&In, _>(got)
        from_readable(got)
    }
}
```

### Current output

```Shell
error[E0275]: overflow evaluating the requirement `for<'a> &'a Wrapper<_>: std::io::Read`
   |
   = help: consider increasing the recursion limit by adding a `#![recursion_limit = "256"]` attribute to your crate (`playground`)
note: required for `&'a Wrapper<Wrapper<_>>` to implement `for<'a> std::io::Read`
  --> src/lib.rs:15:10
   |
15 | impl<In> Read for &Wrapper<In> 
   |          ^^^^     ^^^^^^^^^^^^
16 | where for <'a> &'a In :Read
   |                        ---- unsatisfied trait bound introduced here
   = note: 126 redundant requirements hidden
   = note: required for `&Wrapper<Wrapper<Wrapper<Wrapper<Wrapper<Wrapper<Wrapper<Wrapper<Wrapper<Wrapper<Wrapper<...>>>>>>>>>>>` to implement `for<'a> std::io::Read`
   = note: the full type name has been written to '/playground/target/debug/deps/playground-a9202c2be828afe8.long-type-989942384064675355.txt'

For more information about this error, try `rustc --explain E0275`.
```

### Desired output

```Shell
error[E0282]: type annotations needed
  --> src/lib.rs:38:9
   |
38 |         from_readable(got)
   |         ^^^^^^^^^^^^^ cannot infer type of the type parameter `S` declared on the function `from_readable`
   |
help: consider specifying the generic arguments
   |
38 |         from_readable::<S, Got>(got)
   |                      ++++++++++

For more information about this error, try `rustc --explain E0282`.
error: could not compile `playground` (lib) due to previous error
```

### Rationale and extra context

Current stable channel version 1.74 gives the desired output, but nightly channel version 1.76 gives the misleading infos.

**Comment by mu001999** [comment]
Reduced:
```rust
struct Wrapper<In>(In);

trait Read {}
impl<In> Read for &Wrapper<In> where for<'a> &'a In: Read {}

struct Got;

trait FromReadable<S> {}
impl<S: Read> FromReadable<S> for Got {}

fn from_readable<S, G: FromReadable<S>>(g: G) {}

impl<In> Wrapper<In> {
    fn foo(&self) {
        from_readable(Got)
    }
}
```

**Comment by mu001999** [comment]
91b9ffeab0f8c491c81cd976c0d23d947cd6d12f is the first bad commit.

**Comment by mu001999** [comment]
This code https://github.com/rust-lang/rust/commit/91b9ffeab0f8c491c81cd976c0d23d947cd6d12f#diff-8406bbc0d0b43d84c91b1933305df896ecdba0d1f9269e6744f13d87a2ab268aR429-R441 ensures the `T: WF` obligations come last, so the overflow is emitted first. I think this works with the new trait solver `-Ztrait-solver=next`. cc @estebank
